#QUESTION 1

Word embeddings transform words into vector representations that capture the context and semantics of language. These representations allow similar words to appear close to each other in vector space, making them especially powerful in various NLP applications. Word2Vec is a widely-used embedding method that learns these representations by predicting context words from a target word using simple neural network layers.

Common methods include Word2Vec, GloVe, and FastText.

### Key Benefits of Word Embedding:
1. **Semantic Similarity**: Captures the meaning of words based on their usage.
2. **Efficiency**: Reduces dimensionality compared to sparse representations.
3. **Improved NLP Tasks**: Enhances performance in tasks like sentiment analysis and machine translation.

## Step 1: Removing Stop Words

We start with the original sentence:

"I like Data Science because Data Science is very cool"

Words such as "I" and "is" are considered stop words — they occur frequently but carry little useful meaning. Removing these words helps improve the quality of word embedding models by focusing learning on semantically meaningful tokens.

### After removing stop words, we get:

"like Data Science because Data Science very cool"


## Step 2: Context Window and Training Pairs

We define a context window of size 1, which means for any target word, we consider only one word before and one word after it as context.

For example:
- The word "like" has one context word: "Data".
- The word "Data" appears with "like" and "Science".

These relationships are recorded as (Main, Context) pairs with a label 1 to indicate they are contextually linked. Additionally,there are few negative examples (with label 0) that do not occur together in the sentence to help the model learn to distinguish between related and unrelated word pairs.

### Positive Pairs

| Main     | Context   | Label |
|----------|-----------|-------|
| like     | Data      | 1     |
| Data     | like      | 1     |
| Data     | Science   | 1     |
| Science  | Data      | 1     |
| Science  | because   | 1     |
| because  | Science   | 1     |
| because  | Data      | 1     |
| Data     | because   | 1     |
| Data     | Science   | 1     |
| Science  | Data      | 1     |
| Science  | very      | 1     |
| very     | Science   | 1     |
| very     | cool      | 1     |
| cool     | very      | 1     |

### Negative Pairs

| Main     | Context   | Label |
|----------|-----------|-------|
| like     | Science   | 0     |
| Data     | cool      | 0     |



## Step 3: Initialising Word Vectors

Before training, each word is assigned a random 2D vector. These vectors represent the words both as a main word and a context word. The values are small and randomly chosen.

For example:
- "like" might be [−0.4, 0.2]
- "Data" might be [0.7, −1.0]

These vectors are refined during training so that frequently co-occurring words shift closer together in space, while unrelated words move farther apart.

### Initial Main Vectors

| Word     | X     | Y     |
|----------|--------|--------|
| like     | -0.4  | 0.2    |
| Data     | 0.7   | -1.0   |
| Science  | -0.5  | 0.3    |
| because  | 0.8   | -0.7   |
| very     | -0.9  | 0.6    |
| cool     | 0.4   | -0.9   |

### Initial Context Vectors

| Word     | X     | Y     |
|----------|--------|--------|
| like     | 0.3   | -0.1   |
| Data     | -0.6  | 0.7    |
| Science  | 0.5   | -0.4   |
| because  | -0.8  | 0.9    |
| very     | 0.2   | -0.6   |
| cool     | -0.3  | 0.8    |



## Step 4: Training – Vector Updates Using Online Learning

The Word2Vec model updates vectors by learning to predict context words given a main word.

For each pair:
1. Compute the dot product between main and context vectors.
2. Apply the sigmoid function to map the score to a probability.
3. Calculate the error:  ERROR=LABEL-SCORE
4. Update both vectors using the rule:  
   $$ V_{\text{new}} = V_{\text{old}} + \eta \cdot \text{ERROR} \cdot V_{\text{other}} $$

Where $$ \eta = 0.1 $$ is the learning rate.

If the error is positive, the vectors are moved closer. If negative, they are pushed apart.

### Example: ("like", "Data")

- Main vector (like) = [-0.4, 0.2]  
- Context vector (Data) = [-0.6, 0.7]  
- Dot product = 0.38  
- Sigmoid(0.38) ≈ 0.594  
- Error = 1 − 0.594 = 0.406

**Vector Updates**:
- New like = [-0.424, 0.228]  
- New Data (context) = [-0.616, 0.708]



### Summary of Calculations

| Main     | Context   | Label | Dot Product | Sigmoid | Error  |
|----------|-----------|-------|-------------|---------|--------|
| like     | Data      | 1     | 0.38        | 0.594   | 0.406  |
| like     | Science   | 0     | -0.28       | 0.430   | -0.430 |
| Data     | like      | 1     | 0.31        | 0.577   | 0.423  |
| Data     | Science   | 1     | 0.75        | 0.679   | 0.321  |
| Science  | Data      | 1     | 0.51        | 0.625   | 0.375  |
| Science  | because   | 1     | 0.67        | 0.662   | 0.338  |
| because  | Science   | 1     | 0.68        | 0.664   | 0.336  |
| because  | Data      | 1     | -0.97       | 0.275   | 0.725  |
| Data     | because   | 1     | -1.46       | 0.188   | 0.812  |
| Data     | cool      | 0     | -1.01       | 0.267   | -0.267 |
| Science  | very      | 1     | -0.28       | 0.430   | 0.570  |
| very     | Science   | 1     | -0.69       | 0.334   | 0.666  |
| very     | cool      | 1     | 0.75        | 0.679   | 0.321  |
| cool     | very      | 1     | 0.62        | 0.650   | 0.350  |



## Step 5: Final Vectors After One Pass

### Main Vectors

| Word     | X       | Y       |
|----------|----------|----------|
| like     | -0.359  | 0.178   |
| Data     | 0.749   | -1.070  |
| Science  | -0.451  | 0.265   |
| because  | 0.744   | -0.651  |
| very     | -0.847  | 0.573   |
| cool     | 0.445   | -0.941  |

### Context Vectors

| Word     | X       | Y       |
|----------|----------|----------|
| like     | 0.267   | -0.089  |
| Data     | -0.557  | 0.647   |
| Science  | 0.456   | -0.362  |
| because  | -0.735  | 0.833   |
| very     | 0.184   | -0.555  |
| cool     | -0.267  | 0.720   |


After training, we observe that words like "Data" and "Science" — which appear frequently in similar contexts — have been updated to be closer in vector space. Meanwhile, unrelated pairs such as "Data" and "cool" have moved farther apart. This demonstrates how Word2Vec uses simple mathematical updates to build semantically meaningful representations. These embeddings can now be used in downstream tasks such as similarity search, text classification, and clustering.





# QUESTION 2


## Step 1: Data and Normalisation

We are provided with a small dataset:

| Time Spent (hrs) | Website Visits | Pass/Fail (1/0) |
|------------------|----------------|-----------------|
| 10               | 15             | 1               |
| 15               | 12             | 1               |
| 6                | 5              | 0               |
| 8                | 8              | 0               |

To ensure features are on the same scale, we apply Min-Max Normalisation:

$$
x' = \frac{x - x_{\text{min}}}{x_{\text{max}} - x_{\text{min}}}
$$

- Time Spent Min = 6, Max = 15  
- Website Visits Min = 5, Max = 15

### Normalised Data:

| Time (Norm) | Visits (Norm) | Pass/Fail |
|-------------|----------------|-----------|
| 0.44        | 1              | 1         |
| 1           | 0.7            | 1         |
| 0           | 0              | 0         |
| 0.22        | 0.3            | 0         |



## Step 2: Model Parameters

- **Learning rate (η)**: 0.1  
- **Activation Function**: Step function

$$
g(x) = \begin{cases}
1, & \text{if } x \geq 0 \\
0, & \text{if } x < 0
\end{cases}
$$

- **Initial weights**: $$ w_1 = 0.2, w_2 = 0.3 $$
- **Initial bias**: $$ b = -0.1 $$

The weighted sum (Z) is calculated as:

$$
Z = w_1 \cdot x_1 + w_2 \cdot x_2 + b
$$


## Step 3: First Iteration (Epoch 1)

### Data Point 1:
$$
Z = 0.2 \cdot 0.44 + 0.3 \cdot 1 - 0.1 = 0.288
$$  
Prediction = 1 (Correct) → No update



### Data Point 2:
$$
Z = 0.2 \cdot 1 + 0.3 \cdot 0.7 - 0.1 = 0.31
$$
Prediction = 1 (Correct) → No update


### Data Point 3:
$$
Z = 0.2 \cdot 0 + 0.3 \cdot 0 - 0.1 = -0.1
$$  
Prediction = 0 (Correct) → No update


### Data Point 4:
$$
Z = 0.2 \cdot 0.22 + 0.3 \cdot 0.3 - 0.1 = 0.034
$$
Prediction = 1 (Incorrect)  
True label = 0 → Error = -1

**Update**:
$$
w_1 = 0.2 - 0.1 \cdot 0.22 = 0.1, \quad
w_2 = 0.3 - 0.1 \cdot 0.3 = 0.2, \quad
b = -0.1 - 0.1 = -0.2
$$



## Step 4: Second Iteration (Epoch 2)

### Data Point 1:
$$
Z = 0.1 \cdot 0.44 + 0.2 \cdot 1 - 0.2 = 0.044
$$
Prediction = 1 (Correct) → No update


### Data Point 2:
$$
Z = 0.1 \cdot 1 + 0.2 \cdot 0.7 - 0.2 = 0.14
$$  
Prediction = 1 (Correct) → No update


### Data Point 3:
$$
Z = 0.1 \cdot 0 + 0.2 \cdot 0 - 0.2 = -0.2
$$  
Prediction = 0 (Correct) → No update


### Data Point 4:
$$
Z = 0.1 \cdot 0.22 + 0.2 \cdot 0.3 - 0.2 = -0.018
$$  
Prediction = 0 (Correct) → No update



## Step 5: Final Model

- Final weights: $$ w_1 = 0.1, w_2 = 0.2 $$
- Final bias: $$ b = -0.1 $$

The model has now successfully classified all training examples correctly.


## Step 6: Prediction for New Student

A student studies 12 hours/week and visits the course site 10 times/week.

**Normalised values**:
- Time = $$ \frac{12 - 6}{15 - 6} = 0.667 $$
- Visits = $$ \frac{10 - 5}{15 - 5} = 0.5 $$

**Using final model**:
$$
Z = 0.1 \cdot 0.667 + 0.2 \cdot 0.5 - 0.1 = 0.0667
$$  
Prediction = 1 (Pass)




# QUESTION 3



## Dataset

| Year | Average House Price (million $) |
|------|----------------------------------|
| 1    | 1.1                              |
| 2    | 1.2                              |
| 3    | 1.0                              |



## Recurrent Neural Network (RNN)

**Formulas**:
$$
h_t = \sigma(W_h \cdot h_{t-1} + W_x \cdot x_t + b), \quad y_t = W_y \cdot h_t + b
$$

**Assumptions**:
- Identity Fuction: $$ \sigma(x) = x $$
- Learning rate: 0.01
- Initial hidden state: $$ h_{-1} = 0 $$
- Weights: $$ W_h = 0.5, W_x = 1.0, W_y = 1.0, b = 0 $$


### Year 1 (x₀ = 1.1):
$$
h_0 = W_h \cdot h_{-1} + W_x \cdot x_0 + b
$$
$$
h_0 = 0.5 * 0 + 1.0 * 1.1 = 1.1, \quad y_0 = 1.1
$$

### Year 2 (x₁ = 1.2):
$$
h_1 = W_h \cdot h_0 + W_x \cdot x_1 + b
$$
$$
h_1 = 0.5 * 1.1 + 1.0 * 1.2 = 1.75, \quad y_1 = 1.75
$$

### Year 3 (x₂ = 1.0):
$$
h_2 = W_h \cdot h_1 + W_x \cdot x_2 + b
$$
$$
h_2 = 0.5 * 1.75 + 1.0 * 1.0 = 1.875, \quad y_2 = 1.875
$$

### Year 4 Prediction (x₃ = 1.875):
$$
h_3 = W_h \cdot h_2 + W_x \cdot x_3 + b
$$
$$ h_2 = 1.875 $$
$$
h_3 = 0.5 * 1.875 + 1.0 * 1.875 = 2.8125, \quad y_3 = 2.8125
$$

The RNN model uses a simple architecture where each output depends on the current input and the hidden state from the previous time step. The identity activation function is applied, and weights are kept constant (Wh = 0.5, Wx = 1.0, Wy = 1.0), with zero bias and an initial hidden state of 0.

Over three iterations, the hidden state accumulates prior knowledge, and in Year 4 (2025), the model uses the previously predicted value as input to estimate the future price.

The final output from the RNN is 2.8125, indicating a sharp upward trend, as the model strongly reinforces previous values due to its linear accumulation of context without any control mechanism.



### LSTM Model Calculation:
For each time step \( t \):
1. Forget gate:
   $$ f_t = \sigma(W_f \cdot [h_{t-1}, x_t] + b_f) $$
2. Input gate:
   $$ i_t = \sigma(W_i \cdot [h_{t-1}, x_t] + b_i) $$
   $$ \tilde{C}_t = \tanh(W_c \cdot [h_{t-1}, x_t] + b_c) $$
3. Update cell state:
   $$ C_t = f_t \cdot C_{t-1} + i_t \cdot \tilde{C}_t $$
4. Output gate:
   $$ o_t = \sigma(W_o \cdot [h_{t-1}, x_t] + b_o) $$
   $$ h_t = o_t \cdot \tanh(C_t) $$

**Assumptions**:
- All weights = 1, all biases = 0
- Initial state: $$ h_{-1} = 0, C_{-1} = 0 $$
- Activations:
  - $$ \sigma(x) = \frac{1}{1 + e^{-x}} $$
  - $$ \tanh(x) = \frac{e^x - e^{-x}}{e^x + e^{-x}} $$


### Year 1 Calculation:
1. Forget Gate:
   $$ f_1 = \sigma(W_f \cdot h_0 + x_1 + b_f) = \sigma(1 \cdot 0 + 1.1 + 0) = \sigma(1.1) \approx 0.750 $$
2. Input Gate:
   $$ i_1 = \sigma(W_i \cdot h_0 + x_1 + b_i) = \sigma(1 \cdot 0 + 1.1 + 0) = \sigma(1.1) \approx 0.750 $$
3. Candidate Cell State:
   $$ \tilde{C}_1 = \tanh(W_c \cdot h_0 + x_1 + b_c) = \tanh(1 \cdot 0 + 1.1 + 0) = \tanh(1.1) \approx 0.800 $$
4. Cell State:
   $$ C_1 = f_1 \cdot C_0 + i_1 \cdot \tilde{C}_1 = 0.750 \cdot 0 + 0.750 \cdot 0.800 = 0.600 $$
5. Output Gate:
   $$ o_1 = \sigma(W_o \cdot h_0 + x_1 + b_o) = \sigma(1 \cdot 0 + 1.1 + 0) = \sigma(1.1) \approx 0.750 $$
6. Hidden State:
   $$ h_1 = o_1 \cdot \tanh(C_1) = 0.750 \cdot \tanh(0.600) \approx 0.750 \cdot 0.537 = 0.403 $$

### Year 2 Calculation:
1. Forget Gate:
   $$ f_2 = \sigma(W_f \cdot h_1 + x_2 + b_f) = \sigma(1 \cdot 0.403 + 1.2 + 0) = \sigma(1.603) \approx 0.832 $$
2. Input Gate:
   $$ i_2 = \sigma(W_i \cdot h_1 + x_2 + b_i) = \sigma(1 \cdot 0.403 + 1.2 + 0) = \sigma(1.603) \approx 0.832 $$
3. Candidate Cell State:
   $$ \tilde{C}_2 = \tanh(W_c \cdot h_1 + x_2 + b_c) = \tanh(1 \cdot 0.403 + 1.2 + 0) = \tanh(1.603) \approx 0.922 $$
4. Cell State:
   $$ C_2 = f_2 \cdot C_1 + i_2 \cdot \tilde{C}_2 = 0.832 \cdot 0.600 + 0.832 \cdot 0.922 = 0.499 + 0.767 = 1.266 $$
5. Output Gate:
   $$ o_2 = \sigma(W_o \cdot h_1 + x_2 + b_o) = \sigma(1 \cdot 0.403 + 1.2 + 0) = \sigma(1.603) \approx 0.832 $$
6. Hidden State:
   $$ h_2 = o_2 \cdot \tanh(C_2) = 0.832 \cdot \tanh(1.266) \approx 0.832 \cdot 0.853 = 0.710 $$

### Year 3 Calculation:
1. Forget Gate:
   $$ f_3 = \sigma(W_f \cdot h_2 + x_3 + b_f) = \sigma(1 \cdot 0.710 + 1.0 + 0) = \sigma(1.710) \approx 0.846 $$
2. Input Gate:
   $$ i_3 = \sigma(W_i \cdot h_2 + x_3 + b_i) = \sigma(1 \cdot 0.710 + 1.0 + 0) = \sigma(1.710) \approx 0.846 $$
3. Candidate Cell State:
   $$ \tilde{C}_3 = \tanh(W_c \cdot h_2 + x_3 + b_c) = \tanh(1 \cdot 0.710 + 1.0 + 0) = \tanh(1.710) \approx 0.937 $$
4. Cell State:
   $$ C_3 = f_3 \cdot C_2 + i_3 \cdot \tilde{C}_3 = 0.846 \cdot 1.266 + 0.846 \cdot 0.937 = 1.071 + 0.792 = 1.863 $$
5. Output Gate:
   $$ o_3 = \sigma(W_o \cdot h_2 + x_3 + b_o) = \sigma(1 \cdot 0.710 + 1.0 + 0) = \sigma(1.710) \approx 0.846 $$
6. Hidden State:
   $$ h_3 = o_3 \cdot \tanh(C_3) = 0.846 \cdot \tanh(1.863) \approx 0.846 \cdot 0.952 = 0.806 $$

### Prediction for Year 4 (2025):

1. Forget Gate:
   $$ f_4 = \sigma(W_f \cdot h_3 + x_4 + b_f) = \sigma(1 \cdot 0.806 + 1.875 + 0) = \sigma(2.681) \approx 0.936 $$
2. Input Gate:
   $$ i_4 = \sigma(W_i \cdot h_3 + x_4 + b_i) = \sigma(1 \cdot 0.806 + 1.875 + 0) = \sigma(2.681) \approx 0.936 $$
3. Candidate Cell State:
   $$ \tilde{C}_4 = \tanh(W_c \cdot h_3 + x_4 + b_c) = \tanh(1 \cdot 0.806 + 1.875 + 0) = \tanh(2.681) \approx 0.990 $$
4. Cell State:
   $$ C_4 = f_4 \cdot C_3 + i_4 \cdot \tilde{C}_4 = 0.936 \cdot 1.863 + 0.936 \cdot 0.990 = 1.745 + 0.926 = 2.671 $$
5. Output Gate:
   $$ o_4 = \sigma(W_o \cdot h_3 + x_4 + b_o) = \sigma(1 \cdot 0.806 + 1.875 + 0) = \sigma(2.681) \approx 0.936 $$
6. Hidden State:
   $$ h_4 = o_4 \cdot \tanh(C_4) = 0.936 \cdot \tanh(2.671) \approx 0.936 \cdot 0.990 = 0.926 $$


The LSTM model enhances the traditional RNN by introducing memory cells and gates—forget, input, output, and cell state updates—which help regulate what information should be retained or discarded. This avoids issues like vanishing gradients and improves long-term dependency learning.

All weights and biases were set to 1 and 0 respectively, and sigmoid and tanh functions were used for gate activations and state updates. The initial hidden and cell states were both zero. Through controlled gating, the model refines the influence of past values on each step’s prediction.

The final predicted house price for 2025 using the LSTM model is 0.926, which is significantly lower than the RNN output. This shows that LSTM moderated the influence of increasing trends by balancing both past values and the current input more effectively.



## Final Comparison

| Model | 2025 Prediction |
|-------|------------------|
| RNN   | 2.8125 million   |
| LSTM  | 0.926 million    |

### Conclusion:

The RNN model aggressively amplifies past trends due to its lack of gating, which may lead to overestimation in time-series forecasts. In contrast, the LSTM model provides a more conservative and arguably more realistic forecast by controlling the flow of information through gates, preventing runaway accumulation of hidden state values.

While RNN is simpler and faster to implement, the LSTM model is more reliable for long-term predictions involving time series. It accounts for both recent and distant trends, making it better suited for financial forecasting tasks like house price prediction.





#QUESTION 4
## (a) Data Preparation and Cleaning

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Toxic.csv to Toxic.csv


In [ ]:
import pandas as pd
import string
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

# Load dataset
data = pd.read_csv("Toxic.csv")

# Binarize labels: 1 if any toxicity, else 0
data['label'] = (data[['toxic', 'severe_toxic', 'obscene', 'threat', 'insult', 'identity_hate']]
                 .sum(axis=1) > 0).astype(int)

# Clean text
stop_words = set(stopwords.words('english'))
def clean_text(text):
    text = text.lower().translate(str.maketrans('', '', string.punctuation))
    tokens = text.split()
    return ' '.join([word for word in tokens if word not in stop_words])

data['cleaned'] = data['comment_text'].astype(str).apply(clean_text)
data[['comment_text', 'cleaned', 'label']].head()

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


,comment_text,cleaned,label
0,Explanation\nWhy the edits made under my usern...,explanation edits made username hardcore metal...,0
1,D'aww! He matches this background colour I'm s...,daww matches background colour im seemingly st...,0
2,"Hey man, I'm really not trying to edit war. It...",hey man im really trying edit war guy constant...,0
3,"""\nMore\nI can't make any real suggestions on ...",cant make real suggestions improvement wondere...,0
4,"You, sir, are my hero. Any chance you remember...",sir hero chance remember page thats,0


**Overview:**
The original dataset contains 6 binary labels representing different forms of toxicity. To simplify the task into a binary classification problem, a new label is created where a comment is marked **1 (toxic)** if it contains any of the six labels, and **0 (non-toxic)** otherwise.

**Text cleaning steps include:**
- Converting text to lowercase
- Removing punctuation
- Tokenizing with .split()
- Removing stop words using NLTK's corpus

These cleaned texts are stored in a new column cleaned.

## (b) Word2Vec Embedding and FNN Training

In [ ]:
!pip install --upgrade --force-reinstall numpy gensim

  Using cached numpy-2.2.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)
  Using cached gensim-4.3.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.1 kB)
  Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
  Using cached scipy-1.13.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)
  Using cached smart_open-7.1.0-py3-none-any.whl.metadata (24 kB)
  Using cached wrapt-1.17.2-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.4 kB)
Using cached gensim-4.3.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (26.7 MB)
Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)
Using cached scipy-1.13.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (38.6 MB)
Using cached smart_open-7.1.0-py3-none-any.whl (61 kB)
Using cached wrapt-1.17.2-cp311-cp

In [ ]:
from gensim.models import Word2Vec
import numpy as np

# Train Word2Vec
sentences = [text.split() for text in data['cleaned']]
word2vec_model = Word2Vec(sentences, vector_size=100, window=5, min_count=5, workers=4)

# Create embedding for each comment
def get_embedding(comment):
    words = comment.split()
    valid_words = [word for word in words if word in word2vec_model.wv]
    if valid_words:
        return np.mean([word2vec_model.wv[word] for word in valid_words], axis=0)
    return np.zeros(100)

data['embedding'] = data['cleaned'].apply(get_embedding)
X = np.vstack(data['embedding'].values)
y = data['label'].values

# FNN
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
# Train FNN with Word2Vec embeddings
fnn = Sequential([
    Dense(64, activation='relu', input_shape=(100,)),
    Dense(64, activation='relu'),
    Dense(1, activation='sigmoid')
])
fnn.compile(optimizer=Adam(0.001), loss='binary_crossentropy', metrics=['accuracy'])
fnn.fit(X, y, epochs=5, batch_size=32)

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
4987/4987 ━━━━━━━━━━━━━━━━━━━━ 15s 3ms/step - accuracy: 0.9390 - loss: 0.1657
Epoch 2/5
4987/4987 ━━━━━━━━━━━━━━━━━━━━ 12s 2ms/step - accuracy: 0.9492 - loss: 0.1369
Epoch 3/5
4987/4987 ━━━━━━━━━━━━━━━━━━━━ 12s 2ms/step - accuracy: 0.9493 - loss: 0.1362
Epoch 4/5
4987/4987 ━━━━━━━━━━━━━━━━━━━━ 21s 2ms/step - accuracy: 0.9514 - loss: 0.1338
Epoch 5/5
4987/4987 ━━━━━━━━━━━━━━━━━━━━ 13s 3ms/step - accuracy: 0.9521 - loss: 0.1296


To transform the cleaned comments into numeric representations, a Word2Vec model was trained on the text. Parameters such as vector_size=100, window=5, and min_count=5 were used to control embedding quality and context span. Each comment was then represented by averaging the embeddings of all valid words.

Using these 100-dimensional averaged vectors, a Feedforward Neural Network (FNN) was constructed. The architecture consisted of two dense layers with 64 neurons each, activated by ReLU, and a final sigmoid-activated output layer for binary classification. The model was compiled using binary cross-entropy as the loss function and the Adam optimizer with a learning rate of 0.001. It was trained for five epochs with a batch size of 32.


## (c) Train-Test Split

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

The data is split into 80% training and 20% testing while maintaining label balance. This ensures both sets have a similar distribution of toxic and non-toxic samples.

## (d) Model Building and Training

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Logistic Regression
lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)
y_pred_lr = lr_model.predict(X_test)

In [ ]:
# FNN using train-test
fnn = Sequential([
    Dense(64, activation='relu', input_shape=(100,)),
    Dense(64, activation='relu'),
    Dense(1, activation='sigmoid')
])
fnn.compile(optimizer=Adam(0.001), loss='binary_crossentropy', metrics=['accuracy'])
fnn.fit(X_train, y_train, epochs=5, batch_size=32)
y_pred_fnn = (fnn.predict(X_test) > 0.5).astype(int)

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 12s 3ms/step - accuracy: 0.9380 - loss: 0.1678
Epoch 2/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 21s 3ms/step - accuracy: 0.9493 - loss: 0.1381
Epoch 3/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.9502 - loss: 0.1350
Epoch 4/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.9512 - loss: 0.1319
Epoch 5/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9517 - loss: 0.1293
998/998 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


In [ ]:
# RNN
from tensorflow.keras.layers import SimpleRNN, LSTM, Input, Attention, Flatten
rnn = Sequential([
    SimpleRNN(64, input_shape=(100,1)),
    Dense(64, activation='relu'),
    Dense(1, activation='sigmoid')
])
rnn.compile(optimizer=Adam(0.001), loss='binary_crossentropy', metrics=['accuracy'])
rnn.fit(X_train.reshape(-1, 100, 1), y_train, epochs=5, batch_size=32)
y_pred_rnn = (rnn.predict(X_test.reshape(-1, 100, 1)) > 0.5).astype(int)

/usr/local/lib/python3.11/dist-packages/keras/src/layers/rnn/rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 87s 21ms/step - accuracy: 0.9222 - loss: 0.2080
Epoch 2/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 89s 22ms/step - accuracy: 0.9292 - loss: 0.1924
Epoch 3/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 79s 20ms/step - accuracy: 0.9293 - loss: 0.1932
Epoch 4/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 82s 20ms/step - accuracy: 0.9301 - loss: 0.1897
Epoch 5/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 83s 21ms/step - accuracy: 0.9236 - loss: 0.2063
998/998 ━━━━━━━━━━━━━━━━━━━━ 8s 8ms/step


In [ ]:
# LSTM
lstm = Sequential([
    LSTM(64, input_shape=(100,1)),
    Dense(64, activation='relu'),
    Dense(1, activation='sigmoid')
])
lstm.compile(optimizer=Adam(0.001), loss='binary_crossentropy', metrics=['accuracy'])
lstm.fit(X_train.reshape(-1, 100, 1), y_train, epochs=5, batch_size=32)
y_pred_lstm = (lstm.predict(X_test.reshape(-1, 100, 1)) > 0.5).astype(int)

Epoch 1/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 181s 45ms/step - accuracy: 0.9149 - loss: 0.2375
Epoch 2/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 203s 45ms/step - accuracy: 0.9312 - loss: 0.1891
Epoch 3/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 203s 45ms/step - accuracy: 0.9332 - loss: 0.1841
Epoch 4/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 204s 46ms/step - accuracy: 0.9351 - loss: 0.1781
Epoch 5/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 201s 45ms/step - accuracy: 0.9375 - loss: 0.1710
998/998 ━━━━━━━━━━━━━━━━━━━━ 16s 16ms/step


In [ ]:
# Attention Mechanism
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input

inputs = Input(shape=(100,1))
attention = Attention()([inputs, inputs])
flat = Flatten()(attention)
dense = Dense(64, activation='relu')(flat)
output = Dense(1, activation='sigmoid')(dense)
attention_model = Model(inputs=inputs, outputs=output)
attention_model.compile(optimizer=Adam(0.001), loss='binary_crossentropy', metrics=['accuracy'])
attention_model.fit(X_train.reshape(-1, 100, 1), y_train, epochs=5, batch_size=32)
y_pred_am = (attention_model.predict(X_test.reshape(-1, 100, 1)) > 0.5).astype(int)

Epoch 1/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 22s 5ms/step - accuracy: 0.9233 - loss: 0.2184
Epoch 2/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 19s 5ms/step - accuracy: 0.9426 - loss: 0.1585
Epoch 3/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 20s 5ms/step - accuracy: 0.9440 - loss: 0.1534
Epoch 4/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 19s 5ms/step - accuracy: 0.9440 - loss: 0.1508
Epoch 5/5
3990/3990 ━━━━━━━━━━━━━━━━━━━━ 23s 5ms/step - accuracy: 0.9457 - loss: 0.1477
998/998 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step


To assess the performance of various machine learning and deep learning models on toxic comment classification, five different models were trained using Word2Vec-embedded input features: Logistic Regression (LR), Feedforward Neural Network (FNN), Recurrent Neural Network (RNN), Long Short-Term Memory (LSTM), and a model with an Attention Mechanism.

Each of these models was built with consistency in training parameters to ensure a fair comparison. The settings included two hidden layers (where applicable), each with 64 neurons and ReLU activation. All models used a single sigmoid neuron in the output layer to perform binary classification. The loss function was binary cross-entropy, the optimizer was Adam with a learning rate of 0.001, and training was conducted for 5 epochs with a batch size of 32. The training and test sets were kept constant across all models with an 80-20 stratified split.

###Feedforward Neural Network (FNN)


The FNN was trained on static 100-dimensional Word2Vec embeddings for each comment. It showed a stable learning curve across all epochs, starting with an accuracy of 93.80% and reaching 95.17% by the fifth epoch. The steady decline in loss from 0.1678 to 0.1293 indicates that the network effectively optimized the weights to separate toxic from non-toxic comments. As expected, FNN demonstrated robust performance due to its ability to capture complex patterns through multiple dense layers, even though it lacks any sequential processing capabilities.

Interpretation: The FNN is well-suited for high-dimensional feature vectors like Word2Vec averages. Despite not modeling word order, its non-linearity allowed it to distinguish between comment types effectively.

###Recurrent Neural Network (RNN)


The RNN model was built using TensorFlow’s SimpleRNN layer to process the embedded data as sequences. It initially achieved 92.22% accuracy but did not show consistent improvement. By the fifth epoch, the accuracy slightly declined to 92.36%, and loss values fluctuated — increasing again in the final epoch to 0.2063. This suggests the RNN suffered from the vanishing gradient problem, which is common in basic RNN architectures when learning long-term dependencies.

Interpretation: RNNs are designed to model temporal dependencies, but the simple RNN structure could not retain long-term information effectively, likely leading to underfitting on complex patterns and reduced generalization.

###Long Short-Term Memory (LSTM)


The LSTM model, an advanced version of RNN, was introduced to overcome the limitations of vanishing gradients by using memory gates. It began training with 91.49% accuracy and steadily improved across epochs, reaching 93.75% by the fifth epoch. The loss reduced from 0.2375 to 0.1710, showing the model’s growing ability to discriminate toxic and non-toxic inputs over time.

Interpretation: LSTM's gated architecture effectively captured long-term dependencies in the sequence data. Its higher training time reflects increased computational complexity, but it delivered better learning and generalization than the simple RNN.

###Attention Mechanism


The attention-based model used TensorFlow’s Attention layer to selectively focus on important parts of the input sequences. This model demonstrated strong and consistent improvement during training, starting at 92.33% and climbing to 94.57% accuracy in just 5 epochs. The loss dropped from 0.2184 to 0.1477, reflecting efficient learning. Compared to the RNN and LSTM, the attention model learned faster and reached a better balance between training speed and accuracy.

Interpretation: Attention helped the model dynamically emphasize relevant tokens in each comment, making it particularly effective at handling variable-length and noisy text data. The model’s efficiency and strong performance make it a competitive choice for toxic comment classification.

###Conclusion


All models were successfully trained using the same embedding vectors and configurations. The FNN and Attention Mechanism emerged as the most effective models in terms of both speed and performance. The FNN benefited from non-linear transformations over dense embeddings, while Attention excelled at focusing on semantically meaningful parts of the text. LSTM followed closely behind, offering a balance between temporal context learning and prediction accuracy. The RNN model underperformed due to its architectural limitations. Logistic Regression, not shown in this log but used earlier, serves as a reliable baseline but lacks the capability to model complex non-linear interactions.

## (e) Evaluation and Comparison

In [ ]:
def evaluate_model(y_true, y_pred):
    return (
        accuracy_score(y_true, y_pred),
        precision_score(y_true, y_pred),
        recall_score(y_true, y_pred),
        f1_score(y_true, y_pred)
    )

models = {
    "Logistic Regression": y_pred_lr,
    "FNN": y_pred_fnn,
    "RNN": y_pred_rnn,
    "LSTM": y_pred_lstm,
    "Attention": y_pred_am
}

for name, preds in models.items():
    acc, prec, rec, f1 = evaluate_model(y_test, preds)
    print(f"{name} => Accuracy: {acc:.4f}, Precision: {prec:.4f}, Recall: {rec:.4f}, F1 Score: {f1:.4f}")

Logistic Regression => Accuracy: 0.9386, Precision: 0.7993, Recall: 0.5288, F1 Score: 0.6365
FNN => Accuracy: 0.9480, Precision: 0.8193, Recall: 0.6274, F1 Score: 0.7106
RNN => Accuracy: 0.9255, Precision: 0.6816, Recall: 0.5008, F1 Score: 0.5774
LSTM => Accuracy: 0.9364, Precision: 0.7927, Recall: 0.5066, F1 Score: 0.6182
Attention => Accuracy: 0.9448, Precision: 0.8220, Recall: 0.5837, F1 Score: 0.6826


After training all five models using the same input features and hyperparameters, their effectiveness was assessed using four key evaluation metrics: accuracy, precision, recall, and F1-score. These metrics provide a comprehensive understanding of how well each model distinguishes between toxic and non-toxic comments.

###Logistic Regression


The logistic regression model, although the simplest of all, delivered a solid performance with an accuracy of 93.86%. It achieved a relatively high precision of 0.7993, indicating that when it predicted a comment to be toxic, it was correct most of the time. However, its recall was significantly lower at 0.5288, which means it failed to identify nearly half of the actual toxic comments. Consequently, the F1-score — which balances precision and recall — stood at 0.6365. This result confirms that while logistic regression is efficient and interpretable, it struggles to detect all toxic cases, which could be critical in real-world moderation tasks.

###Feedforward Neural Network (FNN)

The FNN model showed the strongest overall performance, achieving the highest accuracy at 94.80% and the best F1-score at 0.7106. With a precision of 0.8193 and recall of 0.6274, the model demonstrated a strong ability to minimize both false positives and false negatives. The improvement over logistic regression is attributed to the FNN’s capability to capture complex, non-linear relationships within the dense Word2Vec embeddings. Its consistent learning curve and robust evaluation metrics suggest that FNN is highly suitable for toxic comment classification, especially when both interpretability and predictive power are needed.

###Recurrent Neural Network (RNN)

The basic RNN model showed moderate performance, with an accuracy of 92.55%, precision of 0.6816, and recall of 0.5008. Its F1-score was the lowest among all models at 0.5774, reflecting an imbalance between its ability to identify toxic comments and its reliability in doing so without excessive false alarms. The sequential nature of RNNs was expected to offer advantages in handling text data; however, the simple RNN architecture likely struggled with long sequences and lost important contextual cues, resulting in reduced recall.

###Long Short-Term Memory (LSTM)

The LSTM model performed better than the simple RNN, with an accuracy of 93.64%, precision of 0.7927, and recall of 0.5066. The F1-score improved to 0.6182, demonstrating that the LSTM’s memory gating mechanisms allowed it to better capture long-term dependencies in the comment text. However, its performance still lagged behind the FNN and Attention models. Given the significantly higher training time required by LSTM, the marginal gain over simpler models may not justify its complexity unless sequential learning is explicitly required.

###Attention Mechanism

The attention-based model achieved one of the best performances, second only to FNN, with an accuracy of 94.48% and an F1-score of 0.6826. Its precision (0.8220) was slightly higher than the FNN’s, while recall (0.5837) was a bit lower. This suggests that the model was slightly more conservative in predicting toxicity but was more confident when it did. Attention allowed the model to focus on the most relevant parts of the input sequence, which likely contributed to its higher predictive accuracy and interpretability.

###Conclusion

In conclusion, both the Feedforward Neural Network and the Attention Mechanism emerged as the top-performing models. The FNN offered the highest F1-score, making it well-balanced for both detecting toxicity and avoiding false predictions. The Attention model, on the other hand, provided comparable performance with an added benefit of interpretability due to its ability to highlight key terms in the input sequence.

Logistic Regression remains a competitive baseline given its simplicity, though its lower recall makes it less suitable for applications where detecting toxic content is critical. LSTM offers a solid compromise between sequential learning and performance, while RNN underperformed, indicating its limitations in capturing the nuances of longer or more complex comment structures.

For practical deployment in content moderation systems, FNN or Attention-based models are recommended due to their strong balance of efficiency, effectiveness, and adaptability to high-dimensional text embeddings.

#REFERENCES
1.	Bahdanau, D., Cho, K., & Bengio, Y. (2014). Neural machine translation by jointly learning to align and translate. arXiv preprint arXiv:1409.0473.
2.	Brownlee, J. (2018). Deep learning for time series forecasting: predict the future with MLPs, CNNs and LSTMs in Python. Machine Learning Mastery.
3.	Church, K. W. (2017). Word2Vec. Natural Language Engineering, 23(1), 155-162.
4.	Hochreiter, S., & Schmidhuber, J. (1997). Long short-term memory. Neural computation, 9(8), 1735-1780.
5.	Mikolov, T., Chen, K., Corrado, G., & Dean, J. (2013). Efficient estimation of word representations in vector space. arXiv preprint arXiv:1301.3781.
6.	Pedregosa, F., Varoquaux, G., Gramfort, A., Michel, V., Thirion, B., Grisel, O., Blondel, M., Prettenhofer, P., Weiss, R., & Dubourg, V. (2011). Scikit-learn: Machine learning in Python. the Journal of machine Learning research, 12, 2825-2830.
7.	Rosenblatt, F. (1958). The perceptron: a probabilistic model for information storage and organization in the brain. Psychological review, 65(6), 386.

